tardis/visualization/widgets/custom_abundance.py

```diff
@@ -0,0 +1,1700 @@
+
+        box_apply = ipw.VBox(
+            [
+                ipw.Label(value="Apply abundance(s) to:"),
+                self.rbs_single_apply,
+                ipw.HBox([self.rbs_multi_apply, self.irs_shell_range]),
+            ],
+            layout=ipw.Layout(margin="0 0 15px 50px"),
+        )
+
+        box_features = ipw.VBox([box_norm, help_note])
+        box_abundance = ipw.VBox(
+            [
+                box_apply,
+                ipw.HBox([self.box_editor, box_features]),
+                box_add_element,
+            ]
+        )
+        box_density = self.density_editor.display()
+
+        main_tab = ipw.Tab([box_abundance, box_density])
+        main_tab.set_title(0, "Edit Abundance")
+        main_tab.set_title(1, "Edit Density")
+
+        hint = ipw.HTML(
+            value="<b><font size='3'>Save model as file: </font></b>"
+        )
+        box_output = ipw.VBox(
+            [
+                hint,
+                self.input_i_time_0,
+                ipw.HBox(
+                    [self.input_path, self.btn_output, self.ckb_overwrite]
+                ),
+            ]
+        )
+        self.read_abundance()
+        self.density_editor.read_density()
+
+        return ipw.VBox(
+            [
+                self.tbs_scale,
+                self.fig,
+                box_head,
+                main_tab,
+                box_output,
+                self.error_view,
+            ]
+        )
+
+    @error_view.capture(clear_output=True)
+    def to_csvy(self, path, overwrite):
+        """Output CSVY file on the specified path.
+
+        Parameters
+        ----------
+            path : str
+                Output path.
+            overwrite : bool
+                True if overwriting, False otherwise.
+        """
+        if not path.endswith(".csvy"):
+            path += ".csvy"
+
+        if os.path.exists(path) and not overwrite:
+            raise FileExistsError(
+                "The file already exists. Click the 'overwrite' checkbox to overwrite it."
+            )
+        else:
+            self.write_yaml_portion(path)
+            self.write_csv_portion(path)
+
+    @error_view.capture(clear_output=True)
+    def write_yaml_portion(self, path):
+        """Write the YAML portion of the output file.
+
+        Parameters
+        ----------
+            path : str
+        """
+        name = path.split("/")[-1]
+        d_time_0 = self.density_editor.density_t_0 * u.day
+        i_time_0 = self.input_i_time_0.value * u.day
+        custom_yaml = CustomYAML(
+            name, d_time_0, i_time_0, self.velocity[0], self.velocity[-1]
+        )
+        custom_yaml.create_fields_dict(self.elements)
+
+        with open(path, "w") as f:
+            yaml_output = yaml.dump(custom_yaml, sort_keys=False)
+
+            # Add YAML delimiter
+            splits = yaml_output.split("\n")
+            splits[0] = splits[-1] = YAML_DELIMITER
```

Comment:
not clear what that line does (it might be right - just don't understand it)

Reply:
It replaces the first and the last line with `---` in the CSVY file. 
The original first line is a YAML tag: `!!python/object:tardis.visualization.widgets.custom_abundance.CustomYAML` and the original last line is blank, so I add this line to replace them and add yaml delimiter.